File: ml_c/src/gradient_descent.c
/**
 * @file gradient_descent.c
 * @author Chen Liang
 * @brief Implementation of gradient descent in C
 * @version 0.1
 * @date 2021-05-05
 *
 * @copyright Copyright (c) 2021
 *
 */

#include "gradient_descent.h"

double* gradient_desent(double *x,
                        double *y,
                        double *theta,
                        float alpha,
                        int m,
                        int num_iters)
{
    double sum = 0.0L;
    double *tmp_theta = calloc(2, sizeof(double));

    int i = 0;
    int j = 0;

    for (i = 0; i < num_iters; i++)
    {
        tmp_theta[0] = theta[0];
        tmp_theta[1] = theta[1];

        sum = 0.0L;

        for (j = 0; j < m; j++)
        {
            sum += (tmp_theta[0] * x[j]) - y[j];
        }
        theta[0] = theta[0] - alpha * sum / m;

        sum = 0.0L;

        for (j = 0; j < m; j++)
        {
            sum += ((tmp_theta[1] * x[j]) - y[j]) * x[j];
        }
        theta[1] = theta[1] - alpha * sum / m;
    }

    return theta;
}

File: machine_learning_javascript/src/modules/linear_regression/gradient_descent.mjs
/**
 * Name: gradient_descent.mjs
 * Author: Chen Liang
 * Description: My implementation of Machine Learning formulas
 * 				in JavaScript. This is an exercise, so you may
 * 				want to use a professional library for production.
 * Date: 2021-06-15
 */

export const get_theta = function(X, y, alpha,
    theta, num_iters) {

    let sum, tmp_theta;

    let num_train = y.length;
    let num_feat = theta.length;
    let h_x = [];

    while (num_iters > 0) {
        h_x = [];

        tmp_theta = theta;

        for (i = 0; i < num_train; i++) {
            sum = 0.0;

            for (j = 0; j < num_feat; j++) {
                sum += tmp_theta[j] * X[i][j];
            }

            h_x.push(sum);
        }

        for (j = 0; j < num_feat; j++) {
            sum = 0.0;

            for (i = 0; i < num_train; i++) {
                sum += (h_x[i] - y[i]) * X[i][j];
            }

            theta[j] = tmp_theta[j] - (alpha * sum / num_train);
        }

        num_iters--;
    }

    return theta;
}
